File: crux_kv/shell/typescript/keyValue.ts
/// The shell side of `crux_kv`, as one method per operation.
///
/// Each method's shape is the generated `EffectHandler`'s, so an app's handler
/// holds an instance and delegates in one line:
///
/// ```ts
/// const kv = createLocalStorageKeyValueHandler();
///
/// const handler: EffectHandler = {
///   kvGet: (operation) => kv.get(operation),
///   kvSet: (operation) => kv.set(operation),
/// };
/// ```
///
/// Write an object of your own shape to put the store somewhere else —
/// IndexedDB, a server, `sessionStorage` — or to replace a single method and
/// leave the rest.
///
/// ## What each operation answers with
///
/// `get` answers with what is stored under the key, `valueNone()` when nothing
/// is. `set` and `delete` answer with the value they replaced or removed, which
/// is also `valueNone()` when there was none. Nothing here is an error unless
/// the store itself failed; a missing key is an answer.
export interface KeyValueHandler {
  /// Read the bytes stored under `operation.key`.
  get(operation: Get): Promise<ValueResult>;
  /// Write `operation.value`, answering with the value it replaced.
  set(operation: Set): Promise<ValueResult>;
  /// Remove `operation.key`, answering with the value it removed.
  delete(operation: Delete): Promise<ValueResult>;
  /// Whether `operation.key` is in the store.
  exists(operation: Exists): Promise<BoolResult>;
  /// The keys starting with `operation.prefix`, from `operation.cursor`.
  listKeys(operation: ListKeys): Promise<KeysResult>;
}

/// A `KeyValueHandler` over `localStorage`, storing each value as a JSON array
/// of byte values.
///
/// ## Where it stores things
///
/// `localStorage` is shared with everything else on the origin, so `prefix`
/// keeps this store's keys apart from the rest. It is empty by default, which
/// keeps a key the app already stores readable; give it one — `"myapp."` —
/// before the app has anything worth keeping.
///
/// ## When there is no `localStorage`
///
/// There is none on a server, in a worker, or when the browser has storage
/// turned off. Rather than pretend, every operation then answers
/// `keyValueErrorIo`, so the core sees a store that is not working rather than
/// one that is empty. Pass a `Storage` of your own — `sessionStorage`, or an
/// in-memory stand-in in a test — to decide otherwise.
///
/// @param prefix put in front of every key in the underlying storage.
/// @param storage the storage to use; the global `localStorage` when omitted,
///   read at call time so that a page can install one first.
export function createLocalStorageKeyValueHandler(
  prefix = "",
  storage?: Storage,
): KeyValueHandler {
  const open = (): Storage | undefined => {
    if (storage !== undefined) return storage;
    try {
      return globalThis.localStorage;
    } catch {
      // Reading the property itself throws when storage is blocked.
      return undefined;
    }
  };
  const unavailable = keyValueErrorIo("localStorage is not available");

  const read = (store: Storage, key: string): Value => {
    const stored = store.getItem(prefix + key);
    if (stored === null) return valueNone();
    return valueBytes(JSON.parse(stored) as number[]);
  };

  return {
    async get(operation: Get): Promise<ValueResult> {
      const store = open();
      if (store === undefined) return valueResultErr(unavailable);
      return valueResultOk(read(store, operation.key));
    },

    async set(operation: Set): Promise<ValueResult> {
      const store = open();
      if (store === undefined) return valueResultErr(unavailable);
      const previous = read(store, operation.key);
      try {
        store.setItem(
          prefix + operation.key,
          JSON.stringify(Array.from(operation.value)),
        );
      } catch (error) {
        // Over quota, most likely.
        return valueResultErr(
          keyValueErrorIo(error instanceof Error ? error.message : String(error)),
        );
      }
      return valueResultOk(previous);
    },

    async delete(operation: Delete): Promise<ValueResult> {
      const store = open();
      if (store === undefined) return valueResultErr(unavailable);
      const previous = read(store, operation.key);
      store.removeItem(prefix + operation.key);
      return valueResultOk(previous);
    },

    async exists(operation: Exists): Promise<BoolResult> {
      const store = open();
      if (store === undefined) return boolResultErr(unavailable);
      return boolResultOk(store.getItem(prefix + operation.key) !== null);
    },

    /// One page holds every remaining key, so the answer's cursor is always 0:
    /// `localStorage` is walked in one go anyway, and paging it would only
    /// invite a caller to hold a cursor across a write.
    async listKeys(operation: ListKeys): Promise<KeysResult> {
      const store = open();
      if (store === undefined) return keysResultErr(unavailable);

      const keys: string[] = [];
      for (let i = 0; i < store.length; i++) {
        const stored = store.key(i);
        if (stored === null || !stored.startsWith(prefix)) continue;
        const key = stored.slice(prefix.length);
        if (key.startsWith(operation.prefix)) keys.push(key);
      }
      keys.sort();

      const cursor = Number(operation.cursor);
      if (cursor > keys.length) return keysResultErr(keyValueErrorCursorNotFound());

      return keysResultOk(new KeyPage(keys.slice(cursor), BigInt(0)));
    },
  };
}

/// The plain path: one handler over the global `localStorage`, with no prefix.
export const localStorageKeyValueHandler: KeyValueHandler =
  createLocalStorageKeyValueHandler();

File: crux_kv/shell/swift/KeyValue.swift
import Foundation

/// The shell side of `crux_kv`, as one method per operation.
///
/// Each method's shape is the generated `EffectHandler`'s, so an app's handler
/// holds an instance and delegates in one line:
///
/// ```swift
/// struct MyHandler: EffectHandler {
///     let kv = UserDefaultsKeyValueHandler()
///
///     func kvGet(_ operation: GetValue) async -> ValueResult { await kv.get(operation) }
///     func kvSet(_ operation: SetValue) async -> ValueResult { await kv.set(operation) }
/// }
/// ```
///
/// Conform your own type to it to put the store somewhere else — the keychain,
/// a database, a file — or to replace a single method and leave the rest.
///
/// ## What each operation answers with
///
/// `get` answers with what is stored under the key, `Value.none` when nothing
/// is. `set` and `delete` answer with the value they replaced or removed, which
/// is also `Value.none` when there was none. Nothing here is an error unless
/// the store itself failed; a missing key is an answer.
public protocol KeyValueHandler: Sendable {
    /// Read the bytes stored under `operation.key`.
    func get(_ operation: GetValue) async -> ValueResult
    /// Write `operation.value`, answering with the value it replaced.
    func set(_ operation: SetValue) async -> ValueResult
    /// Remove `operation.key`, answering with the value it removed.
    func delete(_ operation: DeleteValue) async -> ValueResult
    /// Whether `operation.key` is in the store.
    func exists(_ operation: KeyExists) async -> ExistsResult
    /// The keys starting with `operation.prefix`, from `operation.cursor`.
    func listKeys(_ operation: ListKeys) async -> KeysResult
}

/// A `KeyValueHandler` over `UserDefaults`, storing each value as `Data`.
///
/// ## Which defaults, and which domain
///
/// A read goes through the whole search list, as every `UserDefaults` read
/// does, so a key the app already uses for a preference is the same key here.
/// `listKeys`, though, reads back only the one persistent domain this store
/// writes to — `domain` names it — because the search list also holds the
/// global domain and the system's own keys, and none of those is a key the
/// app put here.
///
/// The domain defaults to the main bundle's identifier, which is the domain
/// the standard defaults write to. Pass a suite instead, with
/// `init(suiteName:)`, to keep the store's keys apart from the app's
/// preferences; that initialiser names the domain for you. A `UserDefaults`
/// built by hand and handed to `init(defaults:domain:)` needs its domain
/// named too, or `listKeys` will read a domain it does not write to.
///
/// ## Size
///
/// `UserDefaults` is a preferences store, not a database: it is read into
/// memory whole and written out whole. Documents and caches belong in a file.
/// `@unchecked`, because `UserDefaults` is thread-safe but predates `Sendable`
/// and is not marked as such.
public final class UserDefaultsKeyValueHandler: KeyValueHandler, @unchecked Sendable {
    private let defaults: UserDefaults
    private let domain: String?

    /// Stores everything in `defaults`, the standard ones unless told
    /// otherwise, writing to the persistent domain `domain` names.
    ///
    /// - Parameter domain: the domain `defaults` writes to, and the one
    ///   `listKeys` reads back. The main bundle's identifier is the standard
    ///   defaults' own domain, and is the default here; a process without a
    ///   bundle identifier has none, and `listKeys` then answers with an
    ///   error rather than with the whole search list.
    public init(defaults: UserDefaults = .standard, domain: String? = Bundle.main.bundleIdentifier) {
        self.defaults = defaults
        self.domain = domain
    }

    /// Stores everything in the suite `suiteName` names, creating it if it does
    /// not exist — the standard defaults if it cannot be opened.
    public convenience init(suiteName: String) {
        self.init(defaults: UserDefaults(suiteName: suiteName) ?? .standard, domain: suiteName)
    }

    public func get(_ operation: GetValue) async -> ValueResult {
        .ok(Self.value(defaults.data(forKey: operation.key)))
    }

    public func set(_ operation: SetValue) async -> ValueResult {
        let previous = defaults.data(forKey: operation.key)
        defaults.set(Data(operation.value), forKey: operation.key)
        return .ok(Self.value(previous))
    }

    public func delete(_ operation: DeleteValue) async -> ValueResult {
        let previous = defaults.data(forKey: operation.key)
        defaults.removeObject(forKey: operation.key)
        return .ok(Self.value(previous))
    }

    /// `data(forKey:)`, not `object(forKey:)`, so that this agrees with `get`:
    /// a key whose value is not `Data` — one of the system's, say — is not a
    /// key this store holds, and `get` already answers `.none` for it.
    public func exists(_ operation: KeyExists) async -> ExistsResult {
        .ok(defaults.data(forKey: operation.key) != nil)
    }

    /// One page holds every remaining key, so the answer's cursor is always 0:
    /// the domain is small enough to sort in one go, and paging it would only
    /// invite a caller to hold a cursor across a write.
    ///
    /// The keys come from the store's own persistent domain, not from
    /// `dictionaryRepresentation()`, which is the union of the search list and
    /// would answer with the global domain's keys — `AppleLanguages` and the
    /// rest — none of which the app ever wrote here.
    public func listKeys(_ operation: ListKeys) async -> KeysResult {
        guard let domain else {
            return .err(
                .other(
                    message: """
                        listKeys needs the name of the domain the store writes to, \
                        and this process has no bundle identifier to fall back on: \
                        pass `domain:`, or use `init(suiteName:)`.
                        """
                )
            )
        }

        let keys = (defaults.persistentDomain(forName: domain) ?? [:]).keys
            .filter { $0.hasPrefix(operation.prefix) }
            .sorted()

        guard operation.cursor <= UInt64(keys.count) else {
            return .err(.cursorNotFound)
        }

        return .ok(KeyPage(keys: Array(keys[Int(operation.cursor)...]), nextCursor: 0))
    }

    private static func value(_ data: Data?) -> Value {
        guard let data else { return .none }
        return .bytes([UInt8](data))
    }
}
